### ISSUE TITLE
Average loss calculated by pl is different from manual average loss

### ISSUE THREAD

**Issue by dreamgonfly** [PyTorchLightning/pytorch-lightning#2993]
## 🐛 Bug
Validation loss reported by pytorch-lightning on progress bar is different from manually averaged validation loss.

I'm not sure if it's a bug or an intended behavior, but I cannot understand why two values are different.

### To Reproduce

1. Run the below code
2. Observe printed average val loss

![image](https://user-images.githubusercontent.com/2340721/90324307-f1f32e00-dfa7-11ea-92d6-6549943a7b43.png)

1 epoch:
val loss on progress bar: 0.962
manually averaged val loss: 1.2909

2 epoch:
val loss on progress bar: 0.946
manually averaged val loss: 1.2527

#### Code sample
```python
import os

import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST
from torchvision import transforms
import pytorch_lightning as pl


class MNISTModel(pl.LightningModule):
    def __init__(self):
        super(MNISTModel, self).__init__()
        # not the best model...
        self.l1 = torch.nn.Linear(28 * 28, 10)
        self.val_losses = []

    def forward(self, x):
        # called with self(x)
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_nb):
        # REQUIRED
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        tensorboard_logs = {"train_loss": loss}
        return {"loss": loss, "log": tensorboard_logs}

    def validation_step(self, batch, batch_nb):
        # OPTIONAL
        x, y = batch
        y_hat = self(x)
        val_loss = F.cross_entropy(y_hat, y)
        self.val_losses.append(val_loss.item())
        return {
            "loss": val_loss,
            "progress_bar": {"val_loss": val_loss},
            "log": {"val_loss": val_loss},
        }

    def on_validation_epoch_end(self):
        print("average val loss", sum(self.val_losses) / len(self.val_losses))
        self.val_losses = []

    def configure_optimizers(self):
        # REQUIRED
        # can return multiple optimizers and learning_rate schedulers
        # (LBFGS it is automatically supported, no need for closure function)
        return torch.optim.Adam(self.parameters(), lr=0.02)

    def train_dataloader(self):
        # REQUIRED
        return DataLoader(
            MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor()),
            batch_size=32,
        )

    def val_dataloader(self):
        # OPTIONAL
        return DataLoader(
            MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor()),
            batch_size=32,
        )


mnist_model = MNISTModel()

# most basic trainer, uses good defaults (1 gpu)
trainer = pl.Trainer(gpus=1)
trainer.fit(mnist_model)

```
### Expected behavior

val losses manually calculated should match the one reported by progress bar of pytorch-lightning

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version: 1.6
 - OS (e.g., Linux): Base Docker image: nvidia/cuda:10.2-cudnn7-devel-ubuntu18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): None
 - Python version: 3.8
 - CUDA/cuDNN version: CUDA 10.2, cudnn 7
 - GPU models and configuration: Titan RTX
 - Any other relevant information:

**Comment by ydcjeff** [comment]
If I am not mistaken, the `val_loss` you are logging is **per batch** while `avg_val_loss` is **per epoch**

**Comment by awaelchli** [comment]
progress_bar and log dict keys in validation_step are not valid. 
What's happening here is that at the end of the validation epoch, it will display the last val_loss value that your validation_step returned. 
Solution: do not return progress_bar key in validation_step result. Compute the avg_val_loss in the epoch_end method or even better, use the [Results ](https://pytorch-lightning.readthedocs.io/en/latest/results.html)object to average the metric, then you don't have to write epoch_end.

I'm removing the bug label, I don't think this is a bug.

**Comment by Data-drone** [comment]
@awaelchli I just came across this and your results link goes to a 404?

**Comment by awaelchli** [comment]
The Results API does not exist anymore, it got removed. Use the new logging api instead:

https://pytorch-lightning.readthedocs.io/en/stable/logging.html#logging-from-a-lightningmodule

**Comment by AjeyPaiK** [comment]
Even this link is returning a 404. Logging the parameters is confusing indeed in ptl. Please help.

**Comment by akihironitta** [comment]
@AjeyPaiK The related doc pages are:
- https://pytorch-lightning.readthedocs.io/en/1.5.9/extensions/logging.html
- https://pytorch-lightning.readthedocs.io/en/1.5.9/api/pytorch_lightning.core.lightning.html#pytorch_lightning.core.lightning.LightningModule.log

This issue is quite outdated, so if anyone has further questions on this, please create a new GitHub discussion at https://github.com/PyTorchLightning/pytorch-lightning/discussions.

**Comment by evileleven** [comment]
I came across the same question and could do find an answer, also the manually calculated epoch average valid loss is different from the automatic output one.
